src/Compilers/CSharp/Portable/Symbols/LocalSymbol.cs

```diff
@@ -340,6 +340,19 @@ internal virtual bool IsReturnable
             }
         }
 
+        /// <summary>
+        /// When a local variable's type is inferred, it may not be used in the
+        /// expression that computes its value (and type). This property returns
+        /// the expression where a reference to an inferred variable is forbidden.
+        /// </summary>
+        internal virtual SyntaxNode ForbiddenZone => null;
```

Comment:
Nit: consider `DeclaringNode` or `InitializingNode` as alternative names, especially if we also do away with having different diagnostics for each case (`ForbiddenDiagnostic`).